pkg/deploy/server/server.go

```diff
@@ -0,0 +1,367 @@
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+package server
+
+import (
+	"context"
+
+	orgv1 "github.com/eclipse-che/che-operator/pkg/apis/org/v1"
+	"github.com/eclipse-che/che-operator/pkg/deploy"
+	"github.com/eclipse-che/che-operator/pkg/deploy/gateway"
+	"github.com/eclipse-che/che-operator/pkg/util"
+	routev1 "github.com/openshift/api/route/v1"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/api/extensions/v1beta1"
+)
+
+const (
+	AvailableStatus               = "Available"
+	UnavailableStatus             = "Unavailable"
+	RollingUpdateInProgressStatus = "Available: Rolling update in progress"
+)
+
+type Server struct {
+	deployContext *deploy.DeployContext
+	component     string
+}
+
+func NewServer(deployContext *deploy.DeployContext) *Server {
+	return &Server{
+		deployContext: deployContext,
+		component:     deploy.DefaultCheFlavor(deployContext.CheCluster),
+	}
+}
+
+func (s *Server) ExposeCheEndpoint() (bool, error) {
+	done, err := s.DetectDefaultCheHost()
+	if !done {
+		return false, err
+	}
+
+	done, err = s.SyncService()
+	if !done {
+		return false, err
+	}
+
+	done, err = s.ExposeEndpoint()
+	if !done {
+		return false, err
+	}
+
+	done, err = s.UpdateCheURL()
+	if !done {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (s *Server) SyncAll() (bool, error) {
+	done, err := s.SyncLegacyConfigMap()
+	if !done {
+		return false, err
+	}
+
+	done, err = s.SyncPVC()
+	if !done {
+		return false, err
+	}
+
+	done, err = s.SyncConfigMap()
+	if !done {
+		return false, err
+	}
+
+	// ensure configmap is created
+	// the version of the object is used in the deployment
+	exists, err := deploy.GetNamespacedObject(s.deployContext, CheConfigMapName, &corev1.ConfigMap{})
+	if !exists {
+		return false, err
+	}
+
+	done, err = s.SyncDeployment()
+	if !done {
+		return false, err
+	}
+
+	done, err = s.UpdateAvailabilityStatus()
+	if !done {
+		return false, err
+	}
+
+	done, err = s.UpdateCheVersion()
+	if !done {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (s *Server) SyncService() (bool, error) {
+	portName := []string{"http"}
+	portNumber := []int32{8080}
+
+	if s.deployContext.CheCluster.Spec.Metrics.Enable {
+		portName = append(portName, "metrics")
+		portNumber = append(portNumber, deploy.DefaultCheMetricsPort)
+	}
+
+	if s.deployContext.CheCluster.Spec.Server.CheDebug == "true" {
+		portName = append(portName, "debug")
+		portNumber = append(portNumber, deploy.DefaultCheDebugPort)
+	}
+
+	spec := deploy.GetServiceSpec(s.deployContext, deploy.CheServiceName, portName, portNumber, s.component)
+	return deploy.Sync(s.deployContext, spec, deploy.ServiceDefaultDiffOpts)
+}
+
+func (s Server) ExposeEndpoint() (bool, error) {
```

Comment:
The same as above (and for methods below), which endpoint? Please think of better name.